packages/typescript/internal/worker/worker_adapter.ts

```diff
@@ -0,0 +1,163 @@
+let createWatchCompilerHost: typeof ts.createWatchCompilerHost;
+
+const formatHost: ts.FormatDiagnosticsHost = {
+  getCanonicalFileName: (path) => path,
+  getCurrentDirectory: ts.sys.getCurrentDirectory,
+  getNewLine: () => ts.sys.newLine,
+};
+
+/**
+ * Prints a diagnostic result for every compiler error or warning.
+ */
+const reportDiagnostic: ts.DiagnosticReporter = (diagnostic) => {
+  worker.log(ts.formatDiagnostic(diagnostic, formatHost));
+};
+
+/**
+ * Prints a diagnostic every time the watch status changes.
+ * This is mainly for messages like "Starting compilation" or "Compilation completed".
+ */
+const reportWatchStatusChanged: ts.WatchStatusReporter = (diagnostic) => {
+  worker.debug(ts.formatDiagnostic(diagnostic, formatHost));
+};
+
+function createWatchProgram(
+  options: ts.CompilerOptions,
+  tsconfigPath: string,
+  setTimeout: ts.System["setTimeout"]
+) {
+  const host = createWatchCompilerHost(
+      tsconfigPath, options, {...ts.sys, setTimeout},
+      ts.createEmitAndSemanticDiagnosticsBuilderProgram, reportDiagnostic,
+      reportWatchStatusChanged);
+
+  // `createWatchProgram` creates an initial program, watches files, and updates
+  // the program over time.
+  return ts.createWatchProgram(host);
+}
+
+/**
+ * Timestamp of the last worker request.
+ */
+let workerRequestTimestamp: number | undefined;
+/**
+ * The typescript compiler in watch mode.
+ */
+let cachedWatchedProgram:
+  | ts.WatchOfConfigFile<ts.EmitAndSemanticDiagnosticsBuilderProgram>
+  | undefined;
+/**
+ * Callback provided by ts.System which should be called at the point at which
+ * file system changes should be consolidated into a new emission from the
+ * watcher.
+ */
+let consolidateChangesCallback: ((...args: any[]) => void) | undefined;
+let cachedWatchProgramArgs: string | undefined;
+
+function getWatchProgram(
+  args: string[]
+): ts.WatchOfConfigFile<ts.EmitAndSemanticDiagnosticsBuilderProgram> {
+  const newWatchArgs = args.join(" ");
+
+  // Check to see if the watch program needs to be updated or if we can re-use the old one.
+  if (
+    cachedWatchedProgram &&
+    cachedWatchProgramArgs &&
+    cachedWatchProgramArgs !== newWatchArgs
+  ) {
+    cachedWatchedProgram.close();
+    cachedWatchedProgram = undefined;
+    cachedWatchProgramArgs = undefined;
+  }
+
+  // If we have not yet created a watch
+  if (!cachedWatchedProgram) {
+    const parsedArgs = ts.parseCommandLine(args);
+    const tsconfigPath = args[args.indexOf("--project") + 1];
+
+    cachedWatchProgramArgs = newWatchArgs;
+    cachedWatchedProgram = createWatchProgram(
+      parsedArgs.options,
+      tsconfigPath,
+      (callback) => {
+        consolidateChangesCallback = callback;
+      }
+    );
+  }
+
+  return cachedWatchedProgram;
+}
+
+async function emitOnce(args: string[]) {
+  const watchProgram = getWatchProgram(args);
+
+  if (consolidateChangesCallback) {
+    consolidateChangesCallback();
+  }
+
+
+    workerRequestTimestamp = Date.now();
```

Comment:
did something go wrong here? I thought this function returns a Promise, if not it shouldn't be async right?

Reply:
Any async function implicitly returns a promise. See the await on line 117 below as an indicator that asynchronous behavior is happening